sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -95,1056 +39,288 @@
-        return new TypeToken<DoFn.InputProvider<InputT>>() {}.where(
-            new TypeParameter<InputT>() {}, in);
-      }
-
-      @Override
-      public boolean isHidden() {
-        return true;
-      }
-    },
-
-    OUTPUT_RECEIVER(
-        Availability.PROCESS_ELEMENT_ONLY, DoFn.OutputReceiver.class, "outputReceiver") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return new TypeToken<DoFn.OutputReceiver<OutputT>>() {}.where(
-            new TypeParameter<OutputT>() {}, out);
-      }
+  private static final Map<Class<?>, DoFnSignature> SIGNATURE_CACHE = new LinkedHashMap<>();
 
-      @Override
-      public boolean isHidden() {
-        return true;
-      }
-    };
-
-    /**
-     * Create a type token representing the given parameter. May use the type token associated
-     * with the input and output types of the {@link DoFn}, depending on the extra
-     * context.
-     */
-    abstract <InputT, OutputT> TypeToken<?> tokenFor(
-        TypeToken<InputT> in, TypeToken<OutputT> out);
-
-    /**
-     * Indicates whether this enum is for testing only, hence should not appear in error messages,
-     * etc. Defaults to {@code false}.
-     */
-    boolean isHidden() {
-      return false;
+  /** @return the {@link DoFnSignature} for the given {@link DoFn}. */
+  static synchronized DoFnSignature getOrParseSignature(
+      @SuppressWarnings("rawtypes") Class<? extends DoFn> fn) {
+    DoFnSignature signature = SIGNATURE_CACHE.get(fn);
+    if (signature != null) {
+      return signature;
     }
 
-    private final Class<?> rawType;
-    private final Availability availability;
-    private final transient MethodDescription method;
-
-    private AdditionalParameter(Availability availability, Class<?> rawType, String method) {
-      this.availability = availability;
-      this.rawType = rawType;
-      try {
-        this.method = new MethodDescription.ForLoadedMethod(
-            ExtraContextFactory.class.getMethod(method));
-      } catch (NoSuchMethodException | SecurityException e) {
-        throw new RuntimeException(
-            "Unable to access method " + method + " on " + ExtraContextFactory.class, e);
-      }
-    }
+    signature = parseSignature(fn);
+    SIGNATURE_CACHE.put(fn, signature);
+    return signature;
   }
 
-  private static final Map<Class<?>, AdditionalParameter> EXTRA_CONTEXTS;
-  private static final Map<Class<?>, AdditionalParameter> EXTRA_PROCESS_CONTEXTS;
-
-  static {
-    ImmutableMap.Builder<Class<?>, AdditionalParameter> everywhereBuilder =
-        ImmutableMap.<Class<?>, AdditionalParameter>builder();
-    ImmutableMap.Builder<Class<?>, AdditionalParameter> processElementBuilder =
-        ImmutableMap.<Class<?>, AdditionalParameter>builder();
+  /** @return the {@link DoFnInvoker} for the given {@link DoFn}. */
+  public static <InputT, OutputT> DoFnInvoker<InputT, OutputT> newByteBuddyInvoker(
```

Comment:
I almost thing `DoFnReflector` should be renamed `DoFnSignatures` since its main purpose is to create a `DoFnSignature`. This is the only method that stands out, and it is just `getOrParseSignature(...)` composed with `newByteBuddyInvoker`.